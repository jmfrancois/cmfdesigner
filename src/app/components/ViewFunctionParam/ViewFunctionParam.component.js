import React from 'react';
import PropTypes from 'prop-types';

function renderProperties(properties) {
	return properties.map(prop => prop.name).join(', ');
}

function ViewFunctionParam(props) {
	if (!props.params) {
		return 'no params';
	}
	const params = props.params.reduce((acc, param) => {
		if (param.type === 'ObjectPattern') {
			acc.push(`{ ${renderProperties(param.properties)} }`);
		}
		if (param.name && param.value) {
			acc.push(`${param.name} = ${param.value}`);
		}
		if (param.type === 'Identifier') {
			acc.push(`${param.name}`);
		}
		if (param.type === 'AssignmentPattern') {
			let defaultValue = '';
			if (param.defaultValue.type === 'ObjectExpression') {
				defaultValue = `{${param.defaultValue.length}}`;
			} else if (param.defaultValue.type === 'Literal') {
				defaultValue = `'${defaultValue.value}'`;
			} else if (param.defaultValue.type === 'ArrayExpression') {
				defaultValue = `[${defaultValue.length}]`;
			} else if (param.defaultValue.value) {
				defaultValue = param.defaultValue.value;
			}
			acc.push(`${param.name} = ${defaultValue},`);
		}
		if (param.type === 'RestElement') {
			acc.push(`...${param.name}`);
		}
		return acc;
	}, []);
	return <span>({params.join(', ')})</span>;
}

ViewFunctionParam.propTypes = {
	params: PropTypes.arrayOf(PropTypes.shape({
		type: PropTypes.string.isRequired,
		name: PropTypes.string,
		value: PropTypes.any,
		defaultValue: PropTypes.shape({
			type: PropTypes.string.isRequired,
			length: PropTypes.number,
			value: PropTypes.any,
		}),
	})),
};
ViewFunctionParam.displayName = 'ViewFunctionParam';
export default ViewFunctionParam;
