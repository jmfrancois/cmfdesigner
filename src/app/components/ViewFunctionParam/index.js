import ViewFunctionParam from './ViewFunctionParam.component';

export default ViewFunctionParam;
