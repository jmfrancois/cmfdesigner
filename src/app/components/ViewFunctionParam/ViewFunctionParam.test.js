import React from 'react';
import { shallow } from 'enzyme';

import ViewFunctionParam from './ViewFunctionParam.component';

describe('ViewFunctionParam', () => {
	it('should render', () => {
		const wrapper = shallow(
			<ViewFunctionParam />
		);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
