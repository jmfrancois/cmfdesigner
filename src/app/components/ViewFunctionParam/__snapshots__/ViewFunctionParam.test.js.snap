// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewFunctionParam should render 1`] = `null`;
